### ISSUE TITLE
Support for PowerPC64 BE and LE

### ISSUE THREAD

**Issue by runlevel5** [Dr-Noob/cpufetch#85]
POWER8, POWER9 and soon POWER10 are quite popular in enterprise environment and IBM has been pushing the OpenPOWER initiatives which makes this platform more affordable for enthusiasts. Products such as Raptor Blackbird or Talos II are some examples. I think it would be great if this CPU is supported.

I could grant you SSH access to my box if it helps

**Comment by samatjain** [comment]
Do the different POWER series have a logo? Other than the first hits which look non-official, I don't see anything else obvious… https://duckduckgo.com/?t=ffab&q=power9+logo&iax=images&ia=images

OpenPOWER however does have a clear logo: https://duckduckgo.com/?t=ffab&q=openpower+logo&iax=images&ia=images

**Comment by Dr-Noob** [comment]
I also think that supporting PowerPC in cpufetch is interesting. The problem is that I didn't have access to a PowerPC machine until a few weeks now, thanks to a github user that allowed me to work on a Talos II (see [this issue](https://github.com/Dr-Noob/cpufetch/issues/59#issuecomment-821876197)).

Unfortunately, at the moment I don't have time to work on this feature, I hope that I will have the time this summer. I will probably start working on it in July if I have the time. Furthermore, I had a quick look weeks ago and it looks like it will be quite hard to give good support for PowerPC, but I will try my best.

- @runlevel5, if you could give me ssh access would be great, because if I can work on different PowerPC machines, testing will be much easier for me. If you wish, please send me an e-mail so we can coordinate the ssh access. You can find my e-mail at https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=cpufetch-git. As I said before, I won't work on this until summer, so no rush.

- @samatjain I also thought that the logo thing would be an issue because as you said there is no clear logo for PowerPC. If you find any conclusion about this, please let me know. Otherwise, I think I will go for something [like this](https://duckduckgo.com/?t=ffab&q=power9+logo&iax=images&ia=images&iai=https%3A%2F%2Fnicma.com%2Fwp-content%2Fuploads%2F2016%2F09%2FIBM-Power9-logo.png)

I will leave this issue opened to keep track here of the PowerPC cpufetch support and if any other user has the same idea, so they can get into this issue instead of opening a new one. I would also appreciate it if @samatjain and @runlevel5 could test the PowerPC cpufetch version whenever it gets ready.

Thanks!

TL;DR: cpufetch will probably support PowerPC in the near future

**Comment by samatjain** [comment]
Re: getting SSH access, there's also https://kernel.cafe/

Re: logo, thinking about it… if Intel and AMD are displayed for x86, it seems like the intent of the tool is to display the manufacturer of the processor, not the ISA. By that logic, for "ppc64" you should see the logo for IBM or Motorola. That's definitely true for x86, and seems true-ish for ARM as well (I'll let @Dr-Noob weigh in).

Dumping some research:


 * [v8's CPU detection](https://chromium.googlesource.com/v8/v8/+/d567f413c7cc6dc858e2c8fa8862b8032cd6c7be/src/base/cpu.cc), using `/proc/self/auxv`
 * [tensorflow's list of processor features](https://github.com/tensorflow/tensorflow/blob/5dcfc51118817f27fad5246812d83e5dccdc5f72/tensorflow/core/platform/platform_strings.h#L201) which get filled in at compile-time by [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/platform_strings_computed.h#L310); I guess we could look at how the compiler, or whatever is setting these flags, is determining whether this stuff is supported
 * [Linux documentation for some Power-specific stuff](https://elixir.bootlin.com/linux/latest/source/arch/powerpc/include/uapi/asm/auxvec.h) exposed by `/proc/self/auxv`: 
 * It'd be nice to get the "Processor Design Revision", as described here: https://forums.raptorcs.com/index.php/topic,61.msg610.html#msg610. In Raptor CS' systems, I think this information is available in their Linux-based BMC (I have a DD2.3 CPU, which Raptor refers to as "v2", see the version "23" in the screenshot below), so we should be able to get it from normal Linux too:

![image](https://user-images.githubusercontent.com/33427/116763546-a5833d80-a9d2-11eb-91fc-79d5891f94aa.png)

@runlevel5 what processor revision do you have, and what does it say there for "version"?

**Comment by Dr-Noob** [comment]
Hi,

I have finished an initial approach of the cpufetch port to PowerPC. I would appreciate it if you guys can give it a shot (just clone the repo or update it with git pull, since the changes are already in the master branch).

Any kind of suggestions and comments are welcome; not only bug reporting but also any request for changes (add new fields to be displayed, remove fields, etc). If you could also paste the output of your machines it would be great!

Here follows some notes about the PowerPC port. Any comments or help regarding this part are also welcome.

- CPU name: x86 and ARM CPUs always have "commercial", easy to remember CPU names that uniquely identify a CPU, but it seems that PowerPC does not follow this philosophy. I think that with the "PART NUMBER" (see @samatjain image), the CPU can be uniquely identified, but it is not a pretty name. Furthermore, it seems that this information can't be obtained at the user level.
- ASCII logo: According to Wikipedia (https://en.wikipedia.org/wiki/POWER9), not only IBM manufactures PowerPC processors, as you already noted. To me, it seems that finding the actual manufacturer of the processor is not trivial at all. I'm always displaying IBM for the logo, and I have not intention to change this unless I find a reliable and feasible solution.
- Peak Performance: I don't know much about PowerPC architecture. I suspect that I'm computing peak performance wrong because I'm assuming that each physical core has 1 SIMD unit, but I'm not sure about that. Here (https://en.wikichip.org/wiki/ibm/microarchitectures/power9#Slice_Design) explains the slice concept in POWER9, explaining that each slice contains a vector unit. I think that I also read elsewhere that each slice corresponds to one thread. If this is true, it means that each core has 4 or 8 SIMD units (depending on if they are SMT4 or SMT8).
- "Processor Design Revision" (suggested by @samatjain): I'm not sure if this is useful, and from what I understand, this only exists for POWER9.
- Linux dependencies: Unfortunately, there are a lot of features that are fetched using Linux. I don't know if PowerPC is commonly used with other operating systems. I guess that they are used under Linux or AIX (can't support the latter because I do not have access to a PowerPC with AIX).

I'm also adding @avollmerhaus and @erkinalp so they can participate in the conversation if they want to. Lastly, special thanks to @avollmerhaus, who gave me access to a PowerPC machine, which has allowed me to develop this port!

**Comment by runlevel5** [comment]
@Dr-Noob great thanks for your work. The logo and POWER9 info shows up correctly on my Raptor Blackbird computer. I did not see the `Processor Design Revision`, wondering if there is any special flag to toggle the display of such information?



<img width="752" alt="Screenshot 2021-08-02 at 9 13 37 am" src="https://user-images.githubusercontent.com/135605/127788152-e6d9cae5-b4a4-4262-9732-8ef758b43289.png">

**Comment by Dr-Noob** [comment]
I have included what is referred to in IBM manuals as "Processor Design Revision" in the `Microarchitecture` field. Any other suggestions or improvements are welcome.

**Comment by runlevel5** [comment]
I would be great if you could add in `Stepping revision` (please refer to https://wiki.raptorcs.com/wiki/POWER9)

**Comment by Dr-Noob** [comment]
Isn't the thing you refer to as `Stepping revision` what I added in commit 7802505? If not, please give an example output of what you refer to.

**Comment by runlevel5** [comment]
Yes, that's exactly it. My is DD 2.3 so I guess `Microarchitecture` should show `POWER9 DD 2.3`

**Comment by MPC7500** [comment]
@Dr-Noob thank you for your work!

You could also add the logos from Motorola, Freescale, NXP and PA Semi. They all have made or make PowerPC CPUs.
For Motorola.
https://en.wikipedia.org/wiki/P.A._Semi
https://en.wikipedia.org/wiki/NXP_Semiconductors
https://en.wikipedia.org/wiki/Freescale_Semiconductor
https://en.wikipedia.org/wiki/Motorola

Motorola is PPC32, only. Dunno if CPUfetch will ever support this architecture.

**Comment by Dr-Noob** [comment]
> Yes, that's exactly it. My is CPU is IBM POWER9 Sforza DD 2.3 so I guess `Microarchitecture` should show `POWER9 DD 2.3`

That is what cpufetch shows, isn't it?

**Comment by Dr-Noob** [comment]
> @Dr-Noob thank you for your work!

You are welcome! I'm happy to know that my program is also interesting and useful for other people.

> You could also add the logos from Motorola, Freescale, NXP and PA Semi. They all have made or make PowerPC CPUs.
> For Motorola.
> https://en.wikipedia.org/wiki/P.A._Semi
> https://en.wikipedia.org/wiki/NXP_Semiconductors
> https://en.wikipedia.org/wiki/Freescale_Semiconductor
> https://en.wikipedia.org/wiki/Motorola
> 
> Motorola is PPC32, only. Dunno if CPUfetch will ever support this architecture.

I have not added those logos because I don't know a reliable way do detect the manufacturer in PowerPC. For a given PowerPC CPU, it may have been manufactured by different companies (e.g: see https://en.wikipedia.org/wiki/PowerPC_7xx#Device_list). If someone knows a way to do so, I will be happy to add such feature.

PS: @MPC7500, can you paste the output of your PowerPC CPU so I can check if the output is ok or not?

**Comment by MPC7500** [comment]
@Dr-Noob thanks for the feedback. Sadly, I don't have these machines in use, only POWER9. But I think we will get feedback about this soon.

**Comment by Dr-Noob** [comment]
> @Dr-Noob thanks for the feedback. Sadly, I don't have these machines in use, only POWER9. But I think we will get feedback about this soon.

Well, the output of a POWER9 is still interesting because the PowerPC version of cpufetch is pretty new, so there may be a few bugs here and there.

**Comment by MPC7500** [comment]
@Dr-Noob I got some feedback from the MacRumors forums:
https://forums.macrumors.com/threads/cpufetch-port-to-powerpc-help-pleasse.2306366/
Maybe you could look at it there.

Here is my cpuinfo:
```
cpu		: POWER9, altivec supported
clock		: 2436.000000MHz
revision	: 2.2 (pvr 004e 1202)
```

**Comment by Dr-Noob** [comment]
> @Dr-Noob I got some feedback from the MacRumors forums:
> https://forums.macrumors.com/threads/cpufetch-port-to-powerpc-help-pleasse.2306366/
> Maybe you could look at it there.
> 
> Here is my cpuinfo:
> 
> ```
> cpu		: POWER9, altivec supported
> clock		: 2436.000000MHz
> revision	: 2.2 (pvr 004e 1202)
> ```
> 
> On page 59 you will find an overview of part numbers:
> https://wiki.raptorcs.com/wiki/File:POWER9_Sforza_DS_v19_16APR2020_pub.pdf
> 
> cpufetch output attached. But I have to stretch the window from 85x25 to 101x25.
> The logo a little bit smaller would be good.
> ![output](https://user-images.githubusercontent.com/70668364/128206130-4941cd40-4efe-4566-b261-7c0e3a9d07eb.png)

**Comment by Dr-Noob** [comment]
> @Dr-Noob I got some feedback from the MacRumors forums:
> https://forums.macrumors.com/threads/cpufetch-port-to-powerpc-help-pleasse.2306366/
> Maybe you could look at it there.

Thank you very much for posting it in macrumors! I think it will be a good thing for the PowerPC port. 

> 
> Here is my cpuinfo:
> 
> ```
> cpu		: POWER9, altivec supported
> clock		: 2436.000000MHz
> revision	: 2.2 (pvr 004e 1202)
> ```
> 
> On page 59 you will find an overview of part numbers:
> https://wiki.raptorcs.com/wiki/File:POWER9_Sforza_DS_v19_16APR2020_pub.pdf
> 
Unfortunately, /proc/cpuinfo does not provide enough information to determine the manufacturer of a processor. pvr (from cpuinfo) provides important information but (to the best of my knowledge) it only allows to detect the microarchitecture of the CPU, not the manufacturer. And the pdf you link is interesting but again not enough because it is only about POWER9 microarchitecture.

> cpufetch output attached. But I have to stretch the window from 85x25 to 101x25.
> The logo a little bit smaller would be good.
> ![output](https://user-images.githubusercontent.com/70668364/128206130-4941cd40-4efe-4566-b261-7c0e3a9d07eb.png)

Nice, thanks! The output looks good. I've seen many people complaining about the logo being too large, I will make all logos shorter in a future release. Thanks for the suggestion.

**Comment by MPC7500** [comment]
Great. 

lshw -class cpu gives me a more detailed output, even the part # (02CY297):

```
  *-cpu:0                   
       description: POWER9, altivec supported
       product: 02CY297
       vendor: IBM
       physical id: 20
       bus info: cpu@0
       version: 2.2 (pvr 004e 1202)
       serial: YA1935039159
       slot: UOPWR.D100011-Node0-Proc0
       size: 2170MHz
       capacity: 3800MHz
       capabilities: performance-monitor cpufreq
       configuration: threads=4
```

Did you also see wicknix post on MacRumors about the compiling error? He is also on Github.

**Comment by Dr-Noob** [comment]
Thanks for the idea, I have just included it. I wanted to add the part number but I didn't know it was possible under Linux. The method I used is a bit hacky but should work on some ppc CPUs.

I saw the post on macrumors, we are trying to figure out the problem.

**Comment by Dr-Noob** [comment]
> @Dr-Noob I got some feedback from the MacRumors forums:
> https://forums.macrumors.com/threads/cpufetch-port-to-powerpc-help-pleasse.2306366/
> Maybe you could look at it there.
> 
> Here is my cpuinfo:
> 
> ```
> cpu		: POWER9, altivec supported
> clock		: 2436.000000MHz
> revision	: 2.2 (pvr 004e 1202)
> ```
> 
> On page 59 you will find an overview of part numbers:
> https://wiki.raptorcs.com/wiki/File:POWER9_Sforza_DS_v19_16APR2020_pub.pdf
> 
> cpufetch output attached. But I have to stretch the window from 85x25 to 101x25.
> The logo a little bit smaller would be good.
> ![output](https://user-images.githubusercontent.com/70668364/128206130-4941cd40-4efe-4566-b261-7c0e3a9d07eb.png)

I've included a shorter IBM logo that is showed (automatically) if the default logo is too long for the terminal width. 

Furthermore, the PowerPC port seems to work properly (at least for the guys who tried it), so we can conclude that it was integrated successfully. I will close this issue soon if nobody has anything else to add, but feel free to open another issue if you find a problem or bug.

**Comment by MPC7500** [comment]
Sorry I had a typing error above ... the standard size of the terminal window is 80 not 85.
Now I have to stretch the window to 82
![grafik](https://user-images.githubusercontent.com/70668364/129378131-2eb130ac-b54a-4c9d-9781-34075bcdd200.png)

The old design I liked more. But it's only me taste.

Ah and instead of "Name" I would call it P/N, Part Number.

**Comment by Dr-Noob** [comment]
> Sorry I had a typing error above ... the standard size of the terminal window is 80 not 85.
> Now I have to stretch the window to 82
> ![grafik](https://user-images.githubusercontent.com/70668364/129378131-2eb130ac-b54a-4c9d-9781-34075bcdd200.png)
> 
> The old design I liked more. But it's only me taste.
> 
> Ah and instead of "Name" I would call it P/N, Part Number.

Well, I didn't think that it should fit in an 80 width terminal, but it is definitely a good idea to do so. I've added a few more features to make the output shorter, so the output can fit in an 80 terminal width with no problem. If the program runs in a bigger terminal, it will automatically adapt the output to fit the width and try to print the output beautifully.

Regarding the IBM logo, I think this is not a matter of taste; the big logo is much better than the small one. But this is the best I can do. When you have a reduced number of characters to print the logo, the output is worse. It's like pictures resolution; it's easier to make a logo look cooler with a width of 60 characters than 40. Recall that **you can always print the bigger, more beautiful logo if your terminal width is big enough.**

Lastly, you are right about changing `Name` to `Part Number`. It's actually more correct to do so. I've updated that too.

Thanks for your advice. I think the ppc version is much better now with all your suggestions!

**Comment by MPC7500** [comment]
Thanks for the explanation. And I have to thank you for supporting architectures others than x86 and ARM. This is not self-evident. Sadly, you don't support sparc64, but as you said, just fork it.

I guess, you could close this issue.

Thank you for all your effort!
![grafik](https://user-images.githubusercontent.com/70668364/129462407-df2520a1-ff39-488b-b29f-dfab0b29ce8c.png)
